server/src/main/java/io/crate/auth/CachingJwkProvider.java

```diff
@@ -0,0 +1,217 @@
+        this.cache = Caffeine.newBuilder()
+            .maximumSize(5)
+            .expireAfter(new Expiry<String, JwkResult>() {
+
+                @Override
+                public long expireAfterCreate(String key, JwkResult value, long currentTime) {
+                    return value.cacheExpirationTime.toNanos();
+                }
+
+                @Override
+                public long expireAfterUpdate(String key, JwkResult value, long currentTime, long currentDuration) {
+                    return currentDuration;
+                }
+
+                @Override
+                public long expireAfterRead(String key, JwkResult value, long currentTime, long currentDuration) {
+                    return currentDuration;
+                }
+            }).build();
+    }
+
+    static URL urlForDomain(String domain) {
+        if (Strings.isNullOrEmpty(domain)) {
+            throw new IllegalArgumentException("A domain is required");
+        }
+
+        if (!domain.startsWith("http")) {
+            domain = "https://" + domain;
+        }
+
+        try {
+            final URI uri = new URI(domain).normalize();
+            return uri.toURL();
+        } catch (MalformedURLException | URISyntaxException e) {
+            throw new IllegalArgumentException("Invalid jwks uri", e);
+        }
+    }
+
+    @Override
+    public Jwk get(String keyId) {
+        return cache.get(keyId, this::getResult).jwk;
+    }
+
+    private JwkResult getResult(String keyId) {
+        final List<Map<String, Object>> keys;
+        final Map<String, List<String>> responseHeaders;
+
+        try {
+            final URLConnection c = (proxy == null) ? this.url.openConnection() : this.url.openConnection(proxy);
+            if (connectTimeout != null) {
+                c.setConnectTimeout(connectTimeout);
+            }
+            if (readTimeout != null) {
+                c.setReadTimeout(readTimeout);
+            }
+
+            for (Map.Entry<String, String> entry : headers.entrySet()) {
+                c.setRequestProperty(entry.getKey(), entry.getValue());
+            }
+
+            try (InputStream inputStream = c.getInputStream()) {
+                Map<String, Object> result = reader.readValue(inputStream);
+                responseHeaders = c.getHeaderFields();
+                //noinspection unchecked
+                keys = (List<Map<String, Object>>) result.get("keys");
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Cannot obtain jwks from url " + url, e);
+        }
+
+        if (keys == null || keys.isEmpty()) {
+            throw new IllegalArgumentException("No keys found in " + url, null);
+        }
+
+        List<Jwk> jwks = new ArrayList<>();
+
+        for (Map<String, Object> values : keys) {
+            jwks.add(Jwk.fromValues(values));
+        }
+
+        Duration ttl = cacheControlMaxAgeFromRequest(responseHeaders);
```

Comment:
Can we avoid creating `responseHeaders` Map and extract single interesting for us header right after `reader.readValue(inputStream)`?

Reply:
👍